lib/Sema/TypeCheckCodeCompletion.cpp

```diff
@@ -586,13 +586,9 @@ TypeChecker::getTypeOfCompletionOperator(DeclContext *DC, Expr *LHS,
     //     (<LHS>)
     //     (code_completion_expr)))
     CodeCompletionExpr dummyRHS(Loc);
-    auto Args = TupleExpr::create(
-        DC->getASTContext(), SourceLoc(), {LHS, &dummyRHS}, {}, {}, SourceLoc(),
-        /*hasTrailingClosure=*/false, /*isImplicit=*/true);
-    BinaryExpr binaryExpr(opExpr, Args, /*isImplicit=*/true);
-
-    return getTypeOfCompletionOperatorImpl(DC, &binaryExpr,
-                                           referencedDecl);
+    auto *binaryExpr = BinaryExpr::create(DC->getASTContext(), LHS, opExpr,
```

Comment:
Oh right, *that place we make expression nodes on the stack*.

Is there any worry about ASTContext heap growth here? My thought is no since this is, what, at most a handful of words per expr...

Reply:
I think it's okay, I mean we already allocate an ASTNode for the `TupleExpr` here, and will allocate other nodes for the earlier call to `resolveDeclRefExpr`, not to mention all the nodes creating while parsing and type-checking. And even during cached code completion, we throw away the `ASTContext` after a fixed number of completions, so the growth should never be unbounded I believe.